src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultImmutableConverter.cs

```diff
@@ -0,0 +1,142 @@
+
+        private const string ImmutableStackTypeName = "System.Collections.Immutable.ImmutableStack";
+        private const string ImmutableStackGenericTypeName = "System.Collections.Immutable.ImmutableStack`1";
+        private const string ImmutableStackGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableStack`1";
+
+        private const string ImmutableQueueTypeName = "System.Collections.Immutable.ImmutableQueue";
+        private const string ImmutableQueueGenericTypeName = "System.Collections.Immutable.ImmutableQueue`1";
+        private const string ImmutableQueueGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableQueue`1";
+
+        private const string ImmutableSortedSetTypeName = "System.Collections.Immutable.ImmutableSortedSet";
+        private const string ImmutableSortedSetGenericTypeName = "System.Collections.Immutable.ImmutableSortedSet`1";
+
+        private const string ImmutableHashSetTypeName = "System.Collections.Immutable.ImmutableHashSet";
+        private const string ImmutableHashSetGenericTypeName = "System.Collections.Immutable.ImmutableHashSet`1";
+        private const string ImmutableSetGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableSet`1";
+
+        internal delegate object ImmutableCreateRangeDelegate<T>(IEnumerable<T> items);
+        internal delegate IEnumerable CreateImmutableCollectionDelegate(string createRangeDelegateKey, IList sourceList);
+
+        private static readonly ConcurrentDictionary<string, CreateImmutableCollectionDelegate> _CreateImmutableCollectionDelegates = new ConcurrentDictionary<string, CreateImmutableCollectionDelegate>();
+
+        private static ConcurrentDictionary<string, object> _createRangeDelegates = new ConcurrentDictionary<string, object>();
+
+        private static readonly ConcurrentDictionary<Type, string> _createRangeDelegateKeys = new ConcurrentDictionary<Type, string>();
+
+        public string GetConstructingTypeName(string immutableCollectionTypeName)
+        {
+            switch (immutableCollectionTypeName)
+            {
+                case ImmutableListGenericTypeName:
+                case ImmutableListGenericInterfaceTypeName:
+                    return ImmutableListTypeName;
+                case ImmutableStackGenericTypeName:
+                case ImmutableStackGenericInterfaceTypeName:
+                    return  ImmutableStackTypeName;
+                case ImmutableQueueGenericTypeName:
+                case ImmutableQueueGenericInterfaceTypeName:
+                    return  ImmutableQueueTypeName;
+                case ImmutableSortedSetGenericTypeName:
+                    return  ImmutableSortedSetTypeName;
+                case ImmutableHashSetGenericTypeName:
+                case ImmutableSetGenericInterfaceTypeName:
+                    return  ImmutableHashSetTypeName;
+                default:
+                    Debug.Fail("Unknown immutable: " + immutableCollectionTypeName);
```

Comment:
Is this really a debug failure case? If there are other collections not here then we just throw - the Debug.Fail doesn't really add anything and can be confusing if someone is using a debug build.